google/cloud/storage/internal/async/insert_object.h

```diff
@@ -0,0 +1,131 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_ASYNC_INSERT_OBJECT_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_ASYNC_INSERT_OBJECT_H
+
+#include "google/cloud/storage/internal/grpc/ctype_cord_workaround.h"
+#include "google/cloud/storage/internal/hash_function.h"
+#include "google/cloud/storage/object_metadata.h"
+#include "google/cloud/future.h"
+#include "google/cloud/internal/async_streaming_write_rpc.h"
+#include "google/cloud/options.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/version.h"
+#include "absl/strings/cord.h"
+#include <google/storage/v2/storage.pb.h>
+#include <memory>
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * Performs a single attempt to insert an object.
+ *
+ * This implements a coroutine to upload an object. It is used in the
+ * implementation of `AsyncClient::InsertObject()`. The payload for the object
+ * is represented by a single `absl::Cord`.
+ *
+ * Using C++20 coroutines we would write this as:
+ *
+ * @code
+ * future<StatusOr<storage::ObjectMetadata>> AsyncObject(
+ *     std::unique_ptr<storage::internal::HashFunction> hash_function,
+ *     std::unique_ptr<StreamingWriteRpc> rpc,
+ *     google::storage::v2::WriteObjectRequest request, absl::Cord data,
+ *     google::cloud::internal::ImmutableOptions options) {
+ *   auto rpc_ok = co_await rpc->Start();
+ *   while (rpc_ok) {
+ *     auto constexpr kMax = static_cast<std::size_t>(
+ *         google::storage::v2::ServiceConstants::MAX_WRITE_CHUNK_BYTES);
+ *     auto size = std::min(data.size(), kMax);
+ *     auto next = data.Subcord(0, size);
+ *     data = data.Subcord(size, data.size());
+ *     auto crc32c = Crc32c(next);
+ *     hash_function.Update(request.offset(), next, crc32c);
+ *     request.mutable_checksummed_data()->set_content(std::move(next));
+ *     request.mutable_checksummed_data()->set_crc32c(crc32c);
+ *     if (data.empty()) {
+ *       // This is the last block, compute full checksums and set flags
+ *       request.set_finish_write(true);
+ *     }
+ *     // Write the data, breaking out of the loop or error.
+ *     rpc_ok = co_await rpc.Write(request, wopt);
+ *     // We need at least one empty Write() for empty objects, only then we
+ *     // can exit the loop.
+ *     if (data.empty()) break;
+ *   }
+ *   auto response = co_await rpc.Finish();
+ *   if (!response) co_return std::move(response).status();
+ *
```

Comment:
Missing code?  Or, as least some indication that there is missing code?

Reply:
Fixed.